Edgar Osvaldo Navarro García A01644488

# Data management using Pandas

**Data management** is a crucial component to statistical analysis and data science work.

This notebook will show you how to import, view, undertand, and manage your data using the [Pandas](http://pandas.pydata.org) data processing library, i.e., the notebook will demonstrates how to read a dataset into Python, and obtain a basic understanding of its content.

Note that **Python** by itself is a general-purpose programming language and does not provide high-level data processing capabilities.  The **Pandas** library was developed to meet this need. **Pandas** is the most popular Python library for data manipulation, and we will use it extensively in this course. **Pandas** provides high-performance, easy-to-use data structures and data analysis tools.

The main data structure that **Pandas** works with is called a **Data Frame**. This is a two-dimensional table of data in which the rows typically
represent cases and the columns represent variables (e.g. data used in this tutorial).  Pandas also has a one-dimensional data structure called a **Series** that we will encounter when accesing a single column of a Data Frame.

Pandas has a variety of functions named `read_xxx` for reading data in different formats.  Right now we will focus on reading `csv` files, which stands for comma-separated values. However the other file formats include `excel`, `json`, and `sql`.

There are many other options to `read_csv` that are very useful.  For example, you would use the option `sep='\t'` instead of the default `sep=','` if the fields of your data file are delimited by tabs instead of commas.  See [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) for the full documentation for `read_csv`.


## Acknowledgments

- The dataset used in this tutorial is from https://www.coursera.org/ from the course "Understanding and Visualizing Data with Python" by University of Michigan


# Importing libraries


In [110]:
# Import the packages that we will be using
import pandas as pd


# Importing data

In [111]:
# Define where you are running the code: colab or local
RunInColab          = False     # (False: no  | True: yes)

# If running in colab:
if RunInColab:
    # Mount your google drive in google colab
    from google.colab import drive
    drive.mount('/content/drive')

    # Find location
    #!pwd
    #!ls
    #!ls "/content/drive/My Drive/Colab Notebooks/MachineLearningWithPython/"

    # Define path del proyecto
    Ruta            = "/content/drive/My Drive/Colab Notebooks/MachineLearningWithPython/"

else:
    # Define path del proyecto
    Ruta            = ""

In [112]:
# url string that hosts our .csv file
url =Ruta + "cartwheel.csv" #Se trabajo en la misma carpeta

# Read the .csv file and store it as a pandas Data Frame
dataframe = pd.read_csv(url)


If we want to print the information about th output object type we would simply type the following: type(df)

In [113]:
print(type(dataframe))

<class 'pandas.core.frame.DataFrame'>


# Exploring the content of the data set

Use the `shape` method to determine the numbers of rows and columns in a data frame. This can be used to confirm that we have actually obtained the data the we are expecting.

Based on what we see below, the data set being read here has $N_r$ rows, corresponding to $N_r$ observations, and $N_c$ columns, corresponding to $N_c$ variables in this particular data file.

In [114]:
print(dataframe.shape) # (filas, columnas)

(52, 12)


In [115]:
print(dataframe.shape[0]) # filas

52


In [116]:
print(dataframe.shape[1]) # columnas

12


If we want to show the entire data frame we would simply write the following:

In [117]:
print(dataframe)

    ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0    1  56.0      F            1       Y             1   62.00      61.0   
1    2  26.0      F            1       Y             1   62.00      60.0   
2    3  33.0      F            1       Y             1   66.00      64.0   
3    4  39.0      F            1       N             0   64.00      63.0   
4    5  27.0      M            2       N             0   73.00      75.0   
5    6  24.0      M            2       N             0   75.00      71.0   
6    7  28.0      M            2       N             0   75.00      76.0   
7    8  22.0      F            1       N             0   65.00      62.0   
8    9  29.0      M            2       Y             1   74.00      73.0   
9   10  33.0      F            1       Y             1   63.00      60.0   
10  11  30.0      M            2       Y             1   69.50      66.0   
11  12  28.0      F            1       Y             1   62.75      58.0   
12  13  25.0

As you can see, we have a 2-Dimensional object where each row is an independent observation and each coloum is a variable.

Now, use the the `head()` function to show the first 5 rows of our data frame

In [118]:
print(dataframe.head())

   ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4  


Also, you can use the the `tail()` function to show the last 5 rows of our data frame

In [119]:
print(dataframe.tail())

    ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
47  48  24.0      M            2       N             0    79.5      75.0   
48  49  28.0      M            2       N             0    77.8      76.0   
49  50  30.0      F            1       N             0    74.6       NaN   
50  51   NaN      M            2       N             0    71.0      70.0   
51  52  27.0      M            2       N             0     NaN      71.5   

    CWDistance Complete  CompleteGroup  Score  
47          82        N            0.0      8  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8  
51         103        Y            1.0     10  


The columns in a Pandas data frame have names, to see the names, use the `columns` method:

To gather more information regarding the data, we can view the column names with the following function:

In [120]:
print(dataframe.columns)

Index(['ID', 'Age', 'Gender', 'GenderGroup', 'Glasses', 'GlassesGroup',
       'Height', 'Wingspan', 'CWDistance', 'Complete', 'CompleteGroup',
       'Score'],
      dtype='object')


Be aware that every variable in a Pandas data frame has a data type.  There are many different data types, but most commonly you will encounter floating point values (real numbers), integers, strings (text), and date/time values.  When Pandas reads a text/csv file, it guesses the data types based on what it sees in the first few rows of the data file.  Usually it selects an appropriate type, but occasionally it does not.  To confirm that the data types are consistent with what the variables represent, inspect the `dtypes` attribute of the data frame.

In [121]:
for col in dataframe.columns:
    print(col + " - " + str(dataframe[col].dtype))

ID - int64
Age - float64
Gender - object
GenderGroup - int64
Glasses - object
GlassesGroup - int64
Height - float64
Wingspan - float64
CWDistance - int64
Complete - object
CompleteGroup - float64
Score - int64


Summary statistics, which include things like the mean, min, and max of the data, can be useful to get a feel for how large some of the variables are and what variables may be the most important.

In [122]:
# Summary statistics for the quantitative variables
for col in dataframe.columns:
    if dataframe[col].dtype == 'int64' or dataframe[col].dtype == 'float64':
        print(col + ": ")
        print(dataframe[col].describe())



ID: 
count    52.000000
mean     26.500000
std      15.154757
min       1.000000
25%      13.750000
50%      26.500000
75%      39.250000
max      52.000000
Name: ID, dtype: float64
Age: 
count    51.000000
mean     28.411765
std       5.755611
min      22.000000
25%      25.000000
50%      27.000000
75%      30.000000
max      56.000000
Name: Age, dtype: float64
GenderGroup: 
count    52.000000
mean      1.500000
std       0.504878
min       1.000000
25%       1.000000
50%       1.500000
75%       2.000000
max       2.000000
Name: GenderGroup, dtype: float64
GlassesGroup: 
count    52.000000
mean      0.500000
std       0.504878
min       0.000000
25%       0.000000
50%       0.500000
75%       1.000000
max       1.000000
Name: GlassesGroup, dtype: float64
Height: 
count    51.000000
mean     68.971569
std       5.303812
min      61.500000
25%      64.500000
50%      69.000000
75%      73.000000
max      79.500000
Name: Height, dtype: float64
Wingspan: 
count    51.000000
mean     67.

In [123]:
# Drop observations with NaN values
dataframe.Age.dropna().describe()
dataframe.Wingspan.dropna().describe()

count    51.000000
mean     67.313725
std       5.624021
min      57.500000
25%      63.000000
50%      66.000000
75%      72.000000
max      76.000000
Name: Wingspan, dtype: float64

It is also possible to get statistics on the entire data frame or a column as follows

- `df.mean()` Returns the mean of all columns
- `df.corr()` Returns the correlation between columns in a data frame
- `df.count()` Returns the number of non-null values in each data frame column
- `df.max()` Returns the highest value in each column
- `df.min()` Returns the lowest value in each column
- `df.median()` Returns the median of each column
- `df.std()` Returns the standard deviation of each column

In [124]:
for col in dataframe.columns:
    if dataframe[col].dtype == 'int64' or dataframe[col].dtype == 'float64':
        print(col + ": ")
        print( "Mean: " + str(dataframe[col].mean()))
        print( "Min: " + str(dataframe[col].min()))
        print( "Max: " + str(dataframe[col].max()))
        print( "Median: " + str(dataframe[col].median()))
        print( "Std: " + str(dataframe[col].std()))
print(dataframe.count())

ID: 
Mean: 26.5
Min: 1
Max: 52
Median: 26.5
Std: 15.154757228892407
Age: 
Mean: 28.41176470588235
Min: 22.0
Max: 56.0
Median: 27.0
Std: 5.755611072990375
GenderGroup: 
Mean: 1.5
Min: 1
Max: 2
Median: 1.5
Std: 0.5048781642974013
GlassesGroup: 
Mean: 0.5
Min: 0
Max: 1
Median: 0.5
Std: 0.5048781642974013
Height: 
Mean: 68.97156862745098
Min: 61.5
Max: 79.5
Median: 69.0
Std: 5.303812354353807
Wingspan: 
Mean: 67.31372549019608
Min: 57.5
Max: 76.0
Median: 66.0
Std: 5.624020611905442
CWDistance: 
Mean: 85.57692307692308
Min: 63
Max: 115
Median: 85.0
Std: 14.353172982328104
CompleteGroup: 
Mean: 0.8431372549019608
Min: 0.0
Max: 1.0
Median: 1.0
Std: 0.36729002271272454
Score: 
Mean: 7.173076923076923
Min: 2
Max: 10
Median: 8.0
Std: 2.2115663320428895
ID               52
Age              51
Gender           52
GenderGroup      52
Glasses          52
GlassesGroup     52
Height           51
Wingspan         51
CWDistance       52
Complete         52
CompleteGroup    51
Score            52
dtype: 

# How to write a data frame to a File

To save a file with your data simply use the `to_csv` attribute

Examples:
- df.to_csv('myDataFrame.csv')
- df.to_csv('myDataFrame.csv', sep='\t')

In [125]:
dataframe.to_csv('CartwheelDataManagement.csv')

# Rename columns

To change the name of a colum use the `rename` attribute

Example:

df = df.rename(columns={"Age": "Edad"})

df.head()

In [126]:
dataframe = dataframe.rename(columns={"Age": "Edad"})
print(dataframe.columns)

Index(['ID', 'Edad', 'Gender', 'GenderGroup', 'Glasses', 'GlassesGroup',
       'Height', 'Wingspan', 'CWDistance', 'Complete', 'CompleteGroup',
       'Score'],
      dtype='object')


In [127]:
# Back to the original name
dataframe = dataframe.rename(columns={"Edad": "Age"})
print(dataframe.columns)

Index(['ID', 'Age', 'Gender', 'GenderGroup', 'Glasses', 'GlassesGroup',
       'Height', 'Wingspan', 'CWDistance', 'Complete', 'CompleteGroup',
       'Score'],
      dtype='object')


# Selection of colums

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent observations or samples and the columns represent variables.  One common manipulation of a data frame is to extract the data for one case or for one variable.  There are several ways to do this, as shown below.

To extract all the values for one column (variable), use one of the following alternatives.

In [128]:
#a = df.Age
#b = df["Age"]
#c = df.loc[:, "Age"]
d = dataframe.iloc[:, 1]

print(d)

#df[["Gender", "GenderGroup"]]



0     56.0
1     26.0
2     33.0
3     39.0
4     27.0
5     24.0
6     28.0
7     22.0
8     29.0
9     33.0
10    30.0
11    28.0
12    25.0
13    23.0
14    31.0
15    26.0
16    26.0
17    27.0
18    23.0
19    24.0
20    23.0
21    29.0
22    25.0
23    26.0
24    23.0
25    28.0
26    24.0
27    25.0
28    32.0
29    38.0
30    27.0
31    33.0
32    38.0
33    27.0
34    24.0
35    27.0
36    25.0
37    26.0
38    31.0
39    30.0
40    23.0
41    26.0
42    28.0
43    26.0
44    30.0
45    39.0
46    27.0
47    24.0
48    28.0
49    30.0
50     NaN
51    27.0
Name: Age, dtype: float64


# Slicing a data set

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent cases and the columns represent variables.  One common manipulation of a data frame is to extract the data for one observation or for one variable.  There are several ways to do this, as shown below.

Lets say we would like to splice our data frame and select only specific portions of our data.  There are three different ways of doing so.

1. .loc()
2. .iloc()
3. .ix()

We will cover the .loc() and .iloc() splicing functions.


The attibute **.loc()** uses labels/column names, in specific, it takes two single/list/range operator separated by ',', the first one indicates the rows and the second one indicates columns.

In [129]:
# Return all observations of CWDistance
dataframe.loc[:,"CWDistance"]

# Return a subset of observations of CWDistance
dataframe.loc[:9, "CWDistance"]

# Select all rows for multiple columns, ["Gender", "GenderGroup"]
dataframe.loc[:,["Gender", "GenderGroup"]]

# Select multiple columns, ["Gender", "GenderGroup"]me
keep = ['Gender', 'GenderGroup']
df_gender = dataframe[keep]

# Select few rows for multiple columns, ["CWDistance", "Height", "Wingspan"]
dataframe.loc[4:9, ["CWDistance", "Height", "Wingspan"]]

# Select range of rows for all columns
dataframe.loc[10:15,:]



ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
10  11  30.0      M            2       Y             1   69.50      66.0   
11  12  28.0      F            1       Y             1   62.75      58.0   
12  13  25.0      F            1       Y             1   65.00      64.5   
13  14  23.0      F            1       N             0   61.50      57.5   
14  15  31.0      M            2       Y             1   73.00      74.0   
15  16  26.0      M            2       Y             1   71.00      72.0   

    CWDistance Complete  CompleteGroup  Score  
10          96        Y            1.0      6  
11          79        Y            1.0     10  
12          92        Y            1.0      6  
13          66        Y            1.0      4  
14          72        Y            1.0      9  
15         115        Y            1.0      6

The attribute **iloc()** is an integer based slicing.

In [130]:
#Consige todas las filas/entadas de las primeras 4 columnas .
dataframe.iloc[:, :4]


# Consigue las primeras 4 filas/entradas de todas las columnas
dataframe.iloc[:4, :]

# Consigie todas las filas/entradas de las columnas 3-7 (GenderGroup, Glasses, GlassesGroup, Height)
dataframe.iloc[:, 3:7]

# Consigue las filas de la 4 a la 8 y columnas 2 a la 4)
dataframe.iloc[4:8, 2:4]

# This is incorrect: si se quiere seleccionar columnas por nombre no se puede usar iloc se tiene que usar loc
#dataframe.iloc[1:5, ["Gender", "GenderGroup"]]
dataframe.loc[1:5, ["Gender", "GenderGroup"]]

Gender  GenderGroup
1      F            1
2      F            1
3      F            1
4      M            2
5      M            2

# Get unique existing values

List unique values in the one of the columns

df.Gender.unique()


In [131]:
# List unique values in the df['Gender'] column
dataframe.Gender.unique() #Muestra los valores unicos de la columna


array(['F', 'M'], dtype=object)

In [132]:
# Lets explore df["GenderGroup] as well
dataframe.GenderGroup.unique() #Muestra los valores unicos de la columna GenderGroup

array([1, 2], dtype=int64)

# Filter, Sort and Groupby



With **Filter** you can use different conditions to filter columns. For example, df[df[year] > 1984] would give you only the column year is greater than 1984. You can use & (and) or | (or) to add different conditions to your filtering. This is also called boolean filtering.

df[df["Height"] >= 70]

In [133]:
dataframe[dataframe["Height"] > 70] # Filtra las filas donde la altura es mayor a 70

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
4    5  27.0      M            2       N             0    73.0      75.0   
5    6  24.0      M            2       N             0    75.0      71.0   
6    7  28.0      M            2       N             0    75.0      76.0   
8    9  29.0      M            2       Y             1    74.0      73.0   
14  15  31.0      M            2       Y             1    73.0      74.0   
15  16  26.0      M            2       Y             1    71.0      72.0   
21  22  29.0      M            2       N             0    71.0      70.0   
25  26  28.0      M            2       N             0    75.0      76.0   
26  27  24.0      M            2       N             0    78.4      71.0   
27  28  25.0      M            2       Y             1    76.0      73.0   
33  34  27.0      M            2       N             0    77.0      75.0   
38  39  31.0      M            2       Y             1    73.0      74.0   
40  41  23.0      F            1       N             0    70.4      71.0   
41  42  26.0      M            2       Y             1    73.5      72.0   
42  43  28.0      F            1       Y             1    72.5      72.0   
43  44  26.0      F            1       Y             1    72.0      72.0   
46  47  27.0      M            2       N             0    78.0      75.0   
47  48  24.0      M            2       N             0    79.5      75.0   
48  49  28.0      M            2       N             0    77.8      76.0   
49  50  30.0      F            1       N             0    74.6       NaN   
50  51   NaN      M            2       N             0    71.0      70.0   

    CWDistance Complete  CompleteGroup  Score  
4           72        N            0.0      4  
5           81        N            0.0      3  
6          107        Y            1.0     10  
8          106        N            0.0      5  
14          72        Y            1.0      9  
15         115        Y            1.0      6  
21         101        Y            1.0      8  
25         111        Y            1.0     10  
26          92        Y            1.0      7  
27         107        Y            1.0      8  
33         100        Y            1.0      8  
38          72        Y            1.0      9  
40          66        Y            1.0      4  
41         115        Y            1.0      6  
42          81        Y            1.0     10  
43          92        Y            1.0      8  
46          72        N            0.0      7  
47          82        N            0.0      8  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8

With **Sort** is possible to sort values in a certain column in an ascending order using `df.sort_values("ColumnName")` or in descending order using `df.sort_values(ColumnName, ascending=False)`.

Furthermore, it’s possible to sort values by Column1Name in ascending order then Column2Name in descending order by using `df.sort_values([Column1Name,Column2Name],ascending=[True,False])`


df.sort_values("Height")
#df.sort_values("Height",ascending=False)

In [134]:
print(dataframe.sort_values("Height")) # Ordena de menor a mayor
print(dataframe.sort_values("Height", ascending = False)) # Ordena de mayor a menor

    ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
16  17  26.0      F            1       N             0   61.50      59.5   
29  30  38.0      F            1       Y             1   61.50      61.0   
37  38  26.0      F            1       N             0   61.50      59.5   
13  14  23.0      F            1       N             0   61.50      57.5   
0    1  56.0      F            1       Y             1   62.00      61.0   
30  31  27.0      F            1       Y             1   62.00      60.0   
1    2  26.0      F            1       Y             1   62.00      60.0   
11  12  28.0      F            1       Y             1   62.75      58.0   
28  29  32.0      F            1       Y             1   63.00      60.0   
9   10  33.0      F            1       Y             1   63.00      60.0   
32  33  38.0      F            1       N             0   64.00      63.0   
3    4  39.0      F            1       N             0   64.00      63.0   
45  46  39.0

The attribute **Groupby** involves splitting the data into groups based on some criteria, applying a function to each group independently and combining the results into a data structure. df.groupby(col) returns a groupby object for values from one column while df.groupby([col1,col2]) returns a groupby object for values from multiple columns.

df.groupby(['Gender'])

In [135]:

dataframe.groupby(['Gender']) #Se agrupa por genero

Size of each group

df.groupby(['Gender']).size()

df.groupby(['Gender','GenderGroup']).size()

In [136]:
dataframe.groupby(['Gender']).size() #Cuantas entradas hay por genero

Gender
F    26
M    26
dtype: int64

This output indicates that we have two types of combinations.

* Case 1: Gender = F & Gender Group = 1
* Case 2: Gender = M & GenderGroup = 2.  

This validates our initial assumption that these two fields essentially portray the same information.

# Data Cleaning: handle with missing data

Before getting started to work with your data, it's a good practice to observe it thoroughly to identify missing values and handle them accordingly.

When reading a dataset using Pandas, there is a set of values including 'NA', 'NULL', and 'NaN' that are taken by default to represent a missing value.  The full list of default missing value codes is in the '`read_csv`' documentation [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  This document also explains how to change the way that '`read_csv`' decides whether a variable's value is missing.

Pandas has functions called `isnull` and `notnull` that can be used to identify where the missing and non-missing values are located in a data frame.  

Below we use these functions to count the number of missing and non-missing values in each variable of the datasetr.

In [137]:

dataframe.isnull().sum() #Cuantos valores nulos hay en cada columna

ID               0
Age              1
Gender           0
GenderGroup      0
Glasses          0
GlassesGroup     0
Height           1
Wingspan         1
CWDistance       0
Complete         0
CompleteGroup    1
Score            0
dtype: int64

Unfortunately, our output indicates that some of our columns contain missing values so we are no able to continue on doing analysis with those colums

In [138]:
dataframe.isnull().sum()
dataframe.notnull().sum()

ID               52
Age              51
Gender           52
GenderGroup      52
Glasses          52
GlassesGroup     52
Height           51
Wingspan         51
CWDistance       52
Complete         52
CompleteGroup    51
Score            52
dtype: int64

Now we use these functions to count the number of missing and non-missing values in a single variable in the dataset

print( df.Height.notnull().sum() )

print( pd.isnull(df.Height).sum() )

In [139]:
print( pd.isnull(dataframe.Height).sum() )

1


In [140]:
# Extract all non-missing values of one of the columns into a new variable
x = dataframe.Age.dropna().describe() 
x.describe()

count     8.000000
mean     30.645922
std      16.044470
min       5.755611
25%      24.250000
50%      27.705882
75%      35.250000
max      56.000000
Name: Age, dtype: float64

# Add and eliminate columns

In some cases it is useful to create or eiminate new columns

In [141]:
dataframe.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [142]:
# Add a new column with new data
dataframe["ColumnaNueva"] = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21','22','23','24','25','26','27','28','29','30','31','32','33','34','35','36','37','38','39','40','41','42','43','44','45','46','47','48','49','50','51','52']

# Create a column data
NewColumnData = dataframe.Age/dataframe.Age

# Insert that column in the data frame
dataframe.insert(12, "ColumnInserted", NewColumnData, True)

dataframe.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  ColumnInserted ColumnaNueva  
0          79        Y            1.0      7             1.0            1  
1          70        Y            1.0      8             1.0            2  
2          85        Y            1.0      7             1.0            3  
3          87        Y            1.0     10             1.0            4  
4          72        N            0.0      4             1.0            5

In [143]:
# # Eliminate inserted column
dataframe.drop("ColumnInserted", axis=1, inplace = True)
dataframe.drop(columns=['ColumnaNueva'], inplace = True)
# # Remove three columns as index base
#dataframe.drop(dataframe.columns[[12]], axis = 1, inplace = True)
#
dataframe.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [144]:
# # Add new column derived from existing columns
dataframe["Genero"] = dataframe["Gender"] 
# # The new column is a function of another column
dataframe["AgeInMonths"] = dataframe["Age"] * 12
#
dataframe.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score Genero  AgeInMonths  
0          79        Y            1.0      7      F        672.0  
1          70        Y            1.0      8      F        312.0  
2          85        Y            1.0      7      F        396.0  
3          87        Y            1.0     10      F        468.0  
4          72        N            0.0      4      M        324.0

In [145]:
# # Eliminate inserted column
dataframe.drop("AgeInMonths", axis=1, inplace = True)
dataframe.drop("Genero", axis=1, inplace = True)
dataframe.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [146]:
# Add a new column with text labels reflecting the code's meaning

dataframe["GenderGroupNew"] = dataframe.GenderGroup.replace({1: "Female", 2: "Male"})

# Show the first 5 rows of the created data frame
dataframe.head()


ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score GenderGroupNew  
0          79        Y            1.0      7         Female  
1          70        Y            1.0      8         Female  
2          85        Y            1.0      7         Female  
3          87        Y            1.0     10         Female  
4          72        N            0.0      4           Male

In [147]:
## Eliminate inserted column
dataframe.drop("GenderGroupNew", axis=1, inplace = True)
#dataframe.drop(['GenderGroupNew'],vaxis='columns',vinplace=True)
dataframe.head()


ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [148]:
## Add a new column with strata based on these cut points
#
## Create a column data
NewColumnData = dataframe.Age/dataframe.Age
#
## Insert that column in the data frame
dataframe.insert(1, "ColumnStrata", NewColumnData, True)
#
dataframe["ColumnStrata"] = pd.cut(dataframe.Height, [60., 63., 66., 69., 72., 75., 78.])
#
## Show the first 5 rows of the created data frame
dataframe.head()




ID  ColumnStrata   Age Gender  GenderGroup Glasses  GlassesGroup  Height  \
0   1  (60.0, 63.0]  56.0      F            1       Y             1    62.0   
1   2  (60.0, 63.0]  26.0      F            1       Y             1    62.0   
2   3  (63.0, 66.0]  33.0      F            1       Y             1    66.0   
3   4  (63.0, 66.0]  39.0      F            1       N             0    64.0   
4   5  (72.0, 75.0]  27.0      M            2       N             0    73.0   

   Wingspan  CWDistance Complete  CompleteGroup  Score  
0      61.0          79        Y            1.0      7  
1      60.0          70        Y            1.0      8  
2      64.0          85        Y            1.0      7  
3      63.0          87        Y            1.0     10  
4      75.0          72        N            0.0      4

In [149]:
## Eliminate inserted column
dataframe.drop("ColumnStrata", axis=1, inplace = True)
#
dataframe.head()




ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [150]:
# Drop several "unused" columns
vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
dataframe.drop(vars, axis=1, inplace = True)

# Add and eliminate rows

In some cases it is requiered to add new observations (rows) to the data set

In [151]:
# Print tail
dataframe.tail()

Age Gender Glasses  Height  Wingspan  CWDistance Complete  Score
47  24.0      M       N    79.5      75.0          82        N      8
48  28.0      M       N    77.8      76.0          99        Y      9
49  30.0      F       N    74.6       NaN          71        Y      9
50   NaN      M       N    71.0      70.0         101        Y      8
51  27.0      M       N     NaN      71.5         103        Y     10

In [152]:
dataframe.loc[len(dataframe.index)] = [24, 'F', 'Y', 66, 'NaN', 68, 'N', 3]
#
dataframe.tail()




Age Gender Glasses  Height Wingspan  CWDistance Complete  Score
48  28.0      M       N    77.8     76.0          99        Y      9
49  30.0      F       N    74.6      NaN          71        Y      9
50   NaN      M       N    71.0     70.0         101        Y      8
51  27.0      M       N     NaN     71.5         103        Y     10
52  24.0      F       Y    66.0      NaN          68        N      3

In [153]:
## Eliminate inserted row
dataframe.drop([28], inplace = True )
#
dataframe.tail()




Age Gender Glasses  Height Wingspan  CWDistance Complete  Score
48  28.0      M       N    77.8     76.0          99        Y      9
49  30.0      F       N    74.6      NaN          71        Y      9
50   NaN      M       N    71.0     70.0         101        Y      8
51  27.0      M       N     NaN     71.5         103        Y     10
52  24.0      F       Y    66.0      NaN          68        N      3

# Cleaning your data: drop out unused columns and/or drop out rows with any missing values

In [154]:
# Drop unused columns
dataframe.head()
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#dataframe.drop(vars, axis=1, inplace = True)

vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
dataframe = dataframe[vars]

# Drop rows with any missing values
dataframe = dataframe.dropna()

# Drop unused columns and drop rows with any missing values
vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
dataframe = dataframe[vars].dropna()

#df
dataframe.head()

Age Gender Glasses  Height Wingspan  CWDistance Complete  Score
0  56.0      F       Y    62.0     61.0          79        Y      7
1  26.0      F       Y    62.0     60.0          70        Y      8
2  33.0      F       Y    66.0     64.0          85        Y      7
3  39.0      F       N    64.0     63.0          87        Y     10
4  27.0      M       N    73.0     75.0          72        N      4

# Final remarks


- The understanding of your dataset is essential
    - Number of observations
    - Variables
    - Data types: numerical or categorial
    - What are my variables of interest

- There are several ways to do the same thing

- Cleaning your dataset (dropping out rows with any missing values) is a good practice

- The **Pandas** library provides fancy, high-performance, easy-to-use data structures and data analysis tools


# Activity: work with the iris dataset

Repeat this tutorial with the iris data set and respond to the following inquiries

1. Calculate the statistical summary for each quantitative variables. Explain the results
    - Identify the name of each column: Sepal_Length, Sepal_Width, Petal_Length, Petal_Width, Flower
    - Identify the type of each column
    Sepal_Length float64 - continuas
    Sepal_Width float64 - continuas
    Petal_Length float64 - continuas
    Petal_Width float64 - continuas
    Flower int64 - discreta

    - Minimum, maximum, mean, average, median, standar deviation
    Sepal_Length
    -min 4.3
    -max 7.9
    -mean 5.843333
    -average 5.843333
    -median 5.8
    -standar deviation 0.828

    Sepal_Width
    -min 2
    -max 4.4
    -mean 3.054
    -average 3.054
    -median 3
    -standar deviation 0.433594

    Petal_Length
    -min 1
    -max 6.9
    -mean 3.758667
    -average 3.758667
    -median 4.35
    -standar deviation 1.764420

    Petal_Width
    -min 0.1
    -max 2.5
    -mean 1.198667
    -average 1.198667
    -median 1.3
    -standar deviation 0.763161
    
    Flower
    -min 0
    -max 2
    -mean 1
    -average 1
    -median 1
    -standar deviation 0.819232

2. Are there missing data? If so, create a new dataset containing only the rows with the non-missing data
No hay datos perdidos,
print(dataset.isnull().sum())
dataset = dataset.dropna()
print(dataset.shape)

3. Create a new dataset containing only the petal width and length and the type of Flower
newdataframe = dataset[[ "Petal_Length", "Petal_Width", "Flower"]]
newdataframe.head()

4. Create a new dataset containing only the setal width and length and the type of Flower
newdataframe2 = dataset[[ "Sepal_Length", "Sepal_Width", "Flower"]]
newdataframe.head()

5. Create a new dataset containing the setal width and length and the type of Flower encoded as a categorical numerical column
dataset3 = dataset[[ "Sepal_Length", "Sepal_Width", "Flower"]]
dataset3 = dataset3.replace({"Flower":  {"Iris-setosa":0, "Iris-versicolor":1, "Iris-virginica":2} })
print(dataset3)  